js/sri4node-attachments.js

```diff
@@ -1,1078 +1,1677 @@
+        const uploadPromises = fileArrayWithFileObjForUploadHandler.map((file) =>
+          (async () => {
+            try {
+              await runAfterUpload(sriRequest.dbT, sriRequest, file);
+            } catch (ex) {
+              logException(ex, file);
+              throw ex;
+            }
+            debug("handleFile success");
+          })()
+        );
+        await Promise.all(uploadPromises);
       }
-
-      if (bodyJson.some(e => !e.attachment.key)) {
-        throw new sriRequest.SriError({
-          status: 409,
-          errors: [{
-            code: 'missing.json.attachment.key',
-            type: 'ERROR',
-            message: 'each attachment json needs a key'
-          }]
-        });
+    } else {
+      try {
+        await runAfterUpload(sriRequest.dbT, sriRequest, fileArrayWithFileObjForUploadHandler);
+        debug("handleFile success");
+      } catch (ex) {
+        logException(ex, fileArrayWithFileObjForUploadHandler);
+        throw ex;
       }
+    }
+  }
 
-      bodyJson.forEach(att => {
-        if (!att.resource || !att.resource.href) {
-          throw new sriRequest.SriError({
-            status: 409,
-            errors: [{
-              code: 'missing.json.body.resource',
-              type: 'ERROR',
-              message: 'each attachment json needs a resource'
-            }]
-          });
-        } else {
-          let chuncks = att.resource.href.split("/");
-          att.resource.key = chuncks[chuncks.length - 1];
-        }
+  /**
+   * 
+   * @param {*} body 
+   * @param {TSriRequest} sriRequest 
+   */
+  function throwErrorWhenBodyIsMissing(body, sriRequest) {
+    if (body === undefined) {
+      throw new sriRequest.SriError({
+        status: 409,
+        errors: [
+          {
+            code: "missing.body",
+            type: "ERROR",
+            message: "Body is required.",
+          },
+        ],
       });
     }
+  }
 
-    return {
-      customRouteForUpload: function (runAfterUpload, getResourceForCopy) {
-        return {
-          routePostfix: '/attachments',
-          httpMethods: ['POST'],
-          readOnly: false,
-          busBoy: true,
+  /**
+   * This function will rename the uploaded / copied attachments from
+   * their temporarily filenames to their real filenames.
+   * @param {*} attachmentsWithFileObj 
+   * @returns 
+   */
+  async function renameAttachmentsToRealNames(attachmentsWithFileObj) {
+    return Promise.all(attachmentsWithFileObj
+      .map((file) => renameFile(file)));
+  }
 
-          beforeStreamingHandler: async (tx, sriRequest, customMapping) => {
+  /**
+   * 
+   * @param {*} bodyJsonArray 
+   * @returns {Promise<Array<TMultiPartSingleBodyForFileUploads | TBodyForFileCopy>>}
+   */
+  async function convertFilenamesInAttachmentsBodyToSafeFilenames(bodyJsonArray) {
+    return bodyJsonArray.map((b) => ({
+      ...b,
+      file: b.file ? getSafeFilename(b.file) : undefined,
+      originalFilename: b.file,
```

Comment:
Why was this added?

Reply:
to store the original name after getSafeFilename() might have transformed the original filename in an attachment copy